stackslib/src/net/api/gettenure.rs

```diff
@@ -0,0 +1,324 @@
+        let block_stream =
+            NakamotoBlockStream::new(chainstate, block_id, consensus_hash, parent_block_id)?;
+        let headers_conn = chainstate.reopen_db()?;
+        Ok(NakamotoTenureStream {
+            block_stream,
+            headers_conn,
+            total_sent: 0,
+        })
+    }
+
+    /// Start streaming the next block (i.e. the parent of the block we last streamed).
+    /// Return Ok(true) if we can fit the block into the stream.
+    /// Return Ok(false) if not. The caller will need to call this RPC method again with the block
+    /// ID of the last block it received.
+    /// Return Err(..) on DB error
+    pub fn next_block(&mut self) -> Result<bool, ChainError> {
+        let parent_header = NakamotoChainState::get_block_header_nakamoto(
+            &self.headers_conn,
+            &self.block_stream.parent_block_id,
+        )?
+        .ok_or(ChainError::NoSuchBlockError)?;
+
+        // stop sending if the parent is an epoch2 block
+        let Some(parent_nakamoto_header) = parent_header.anchored_header.as_stacks_nakamoto()
+        else {
+            return Ok(false);
+        };
+
+        // stop sending if the parent is in a different tenure
+        if parent_nakamoto_header.consensus_hash != self.block_stream.consensus_hash {
+            return Ok(false);
+        }
+
+        let parent_size = self
+            .block_stream
+            .staging_db_conn
+            .conn()
+            .get_nakamoto_block_size(&self.block_stream.parent_block_id)?
+            .ok_or(ChainError::NoSuchBlockError)?;
+
+        self.total_sent = self
+            .total_sent
+            .saturating_add(self.block_stream.total_bytes);
+        if self.total_sent.saturating_add(parent_size) > MAX_MESSAGE_LEN.into() {
+            // out of space to send this
+            return Ok(false);
+        }
+
+        self.block_stream.reset(
+            parent_nakamoto_header.block_id(),
+            parent_nakamoto_header.parent_block_id.clone(),
+        )?;
+        Ok(true)
+    }
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCNakamotoTenureRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(r#"^/v3/tenures/(?P<block_id>[0-9a-f]{64})$"#).unwrap()
+    }
+
+    /// Try to decode this request.
+    /// There's nothing to load here, so just make sure the request is well-formed.
+    fn try_parse_request(
+        &mut self,
+        preamble: &HttpRequestPreamble,
+        captures: &Captures,
+        query: Option<&str>,
+        _body: &[u8],
+    ) -> Result<HttpRequestContents, Error> {
+        if preamble.get_content_length() != 0 {
+            return Err(Error::DecodeError(
+                "Invalid Http request: expected 0-length body".to_string(),
+            ));
+        }
+
+        let block_id_str = captures
+            .name("block_id")
+            .ok_or(Error::DecodeError(
```

Comment:
Lazy evaluation will avoid unnecessary string allocations:

```suggestion
            .ok_or_else(|| Error::DecodeError(
```